python/mspasspy/util/decorators.py

```diff
@@ -58,9 +72,18 @@ def mspass_func_wrapper(func, data, *args, object_history=False, alg_id=None, al
         res = func(data, *args, **kwargs)
         if object_history:
             logging_helper.info(data, alg_id, alg_name)
-        if res is None and inplace_return:
+        if inplace_return:
             return data
-        return res
+        elif not function_return_key == None:
```

Comment:
It appears that the `function_return_key` should be a independent logic from the `inplace_return`. This  `function_return_key` handling should move up and move out of the if-else statements. Right?

btw, the optimal way of testing not None should be `function_return_key is not None`.

Reply:
Good point but the change I implemented was slightly different.  The previous logic is correct if inplace_return is False.   However, the correct usage in all cases if using the function_return_key feature is that inplace_return is implied to be True.  That is a definitely a mistake and I think the new logic is correct.   

Now before I push these changes, however, I ran into a surprise when I tried to push the changes I made.   First, git push refused with this message:
```
(base) pavlis@pavlis-linux:~/src/mspass/python/mspasspy/util$ git push origin fix_algorithm_error_handling
To github.com:mspass-team/mspass.git
 ! [rejected]        fix_algorithm_error_handling -> fix_algorithm_error_handling (fetch first)
error: failed to push some refs to 'git@github.com:mspass-team/mspass.git'
hint: Updates were rejected because the remote contains work that you do
hint: not have locally. This is usually caused by another repository pushing
hint: to the same ref. You may want to first integrate the remote changes
hint: (e.g., 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
```
So, I did a pull but got this surprising output:
```
(base) pavlis@pavlis-linux:~/src/mspass/python/mspasspy/util$ git pull origin fix_algorithm_error_handling
remote: Enumerating objects: 84, done.
remote: Counting objects: 100% (46/46), done.
remote: Compressing objects: 100% (20/20), done.
remote: Total 84 (delta 30), reused 26 (delta 26), pack-reused 38
Unpacking objects: 100% (84/84), 130.11 KiB | 135.00 KiB/s, done.
From github.com:mspass-team/mspass
 * branch            fix_algorithm_error_handling -> FETCH_HEAD
   ceab2fe..ba089ab  fix_algorithm_error_handling -> origin/fix_algorithm_error_handling
Removing python/tests/testpickle.py
Merge made by the 'recursive' strategy.
 cxx/include/mspass/algorithms/Taper.h |   1 -
 cxx/include/mspass/seismic/Ensemble.h | 140 ++++++++++++++++++++++++
 cxx/python/seismic/seismic_py.cc      | 161 +++++++++++++++++++++++++---
 data/yaml/mspass.yaml                 | 196 ++++++++++++++++++----------------
 data/yaml/mspass_lite.yaml            |  13 ++-
 docs/source/conf.py                   |  11 +-
 python/mspasspy/db/database.py        |  13 ++-
 python/tests/db/test_db.py            |  43 +++++++-
 python/tests/db/test_schema.py        |  16 +--
 python/tests/helper.py                |   1 -
 python/tests/test_ccore.py            | 110 +++++++++++++------
 python/tests/testpickle.py            |  39 -------
 requirements.txt                      |   1 +
 13 files changed, 554 insertions(+), 191 deletions(-)
 delete mode 100644 python/tests/testpickle.py
```
I am very reluctant to push this branch back as I don't know where those changes came from.   Maybe you can insert this code segment in the area under discussion of decorators.py and commit the changes?
```
    try:
        res = func(data, *args, **kwargs)
        if object_history:
            logging_helper.info(data, alg_id, alg_name)
        if function_return_key is not None:
            if isinstance(function_return_key,str):
                data[function_return_key]=res
            else:
                data.elog.log_error(alg_name,
                 "Illegal type received for function_return_key argument="+str(type(function_return_key)+"\nReturn value not saved in Metadata"),
                 ErrorSeverity.Complaint)
            if not inplace_return:
                data.elog.log_error(alg_name,
                  "Inconsistent arguments; inplace_return was set False and function_return_key was not None.\nAssuming inplace_return == True is correct",
                  ErrorSeverity.Complaint)
            return data
        elif inplace_return:
            return data
        else:
            return res
    except RuntimeError as err:
```
where I pasted more than needed to be sure you know the context.

Reply:
As we decided in the call today I committed the changes above and just pushed to this branch.  Hopefully it is ready to merge